straxen/plugins/peaks/peak_subtyping.py

```diff
@@ -0,0 +1,621 @@
+
+        # what else? undefined large peaks are lonepS2
+        lonepS2_indices = (mask == PeakSubtyping.Undefined) & _ps2_mask
+        lonepS2_containers = limited_containers(
+            base_times[lonepS2_indices], after_s2_window_ext, limitation)
+        # assign lonepS2's following peaks
+        tw_after_loneps2 = strax.touching_windows(_peaks, lonepS2_containers)
+        S2PH_indices, S2olS2_indices = combine_indices_ref(
+            tw_after_loneps2, base_areas, large_s2_threshold)
+        mask[S2PH_indices] = PeakSubtyping.S2PH
+        mask[S2olS2_indices] = PeakSubtyping.S2olS2
+
+        # assign fakeS2, pS2, lonepS2
+        mask[fakeS2_indices] = PeakSubtyping.fakeS2
+        mask[pS2_indices] = PeakSubtyping.pS2
+        mask[lonepS2_indices] = PeakSubtyping.lonepS2
+
+        # remaining small S2s are DE
+        undefined_mask = (mask == PeakSubtyping.Undefined)
+        mask[undefined_mask] = PeakSubtyping.DE
+        return mask
+
+    def compute(self, peaks):
+        # Sort the peaks first by center_time
+        peaks = np.sort(peaks, order='center_time')
+
+        result = np.zeros(len(peaks), self.dtype)
+        result['time'] = peaks['time']
+        result['endtime'] = strax.endtime(peaks)
+        result['subtype'] = np.ones(len(peaks)) * PeakSubtyping.Undefined
+        junk_mask = (peaks['type'] != 1) & (peaks['type'] != 2)
+        junk_mask |= np.isnan(peaks['area'])
+        result['subtype'][junk_mask] = PeakSubtyping.Junk
+
+        # mark with S1-pS2 pairing
+        undefined_mask = result['subtype'] == PeakSubtyping.Undefined
+        result['subtype'][undefined_mask] = self.mark_peaks_after_s1s(
+            peaks[undefined_mask],
+            self.subtype_dtype,
+            self.mis_s2_threshold, self.large_s2_threshold, self.other_large_s2_fac,
+            self.s1_s2_window, self.after_s1_window_ext, self.after_s2_window_ext)
+
+        # mark with lonepS2 and fakeS2-pS2 pairing
+        undefined_mask = result['subtype'] == PeakSubtyping.Undefined
+        result['subtype'][undefined_mask] = self.mark_other_s2s(
+            peaks[undefined_mask],
+            self.subtype_dtype,
+            self.large_s2_threshold, self.other_large_s2_fac,
+            self.s1_s2_window, self.after_s2_window_ext)
+
+        # check if there is any undefined peaks
+        n_undefined = (result['subtype'] == PeakSubtyping.Undefined).sum()
+        if n_undefined:
+            raise RuntimeError(f'We still find {n_undefined} peaks undefined!')
+
+        # sort the result by time
+        result = np.sort(result, order='time')
+
+        return result
+
+
+def limited_containers(anchor, length, limitation, forward=True):
+    """
+    Build container given anchor, length and limitation of end time.
+    The retuned containers should not be overlapping
+    :param anchor: sorted array of interval anchor points
+    :param length: length of each interval, len(anchor) == len(length)
+    """
```

Comment:
Can you guys give me an example and explanation what limitation should be? I do not fully understand the purpose of this function.